### ISSUE TITLE
stax.Dropout: Is 'rate' a keep rate or a lose rate?

### ISSUE THREAD

**Issue by cboulay** [google/jax#849]
In Tensorflow, the Dropout layer [`rate` argument](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Dropout#arguments) is the fraction of units to drop.

In stax.Dropout, the [rate argument is passed to random.bernoulli](https://github.com/google/jax/blob/619b033474912d3f0229ddd13d24c5d0af93911c/jax/experimental/stax.py#L286), which returns True for [values **less than** rate](https://github.com/google/jax/blob/619b033474912d3f0229ddd13d24c5d0af93911c/jax/random.py#L426). The array of boolean values is then used in [`np.where`](https://github.com/google/jax/blob/619b033474912d3f0229ddd13d24c5d0af93911c/jax/experimental/stax.py#L287), returning `inputs / rate` when True and `0` when False.

In other words, it seems like `rate` values will be kept and scaled up by 1/rate, and `1-rate` values will be dropped.

Is this opposite-to-Tensorflow usage of `rate` intended? 

```Python
import numpy as onp
from jax import numpy as np
from jax import random
from jax.experimental import stax


DATA_SHAPE = (100, 1)
rng = random.PRNGKey(onp.random.randint(0, 10000))
do_init, do_apply = stax.Dropout(0.2)
out_shape, do_params = do_init(rng, DATA_SHAPE)
data = random.uniform(rng, DATA_SHAPE)

rng, subkey = random.split(rng)
data_w_drops = do_apply(do_params, data, rng=subkey)
print(np.sum(data_w_drops > 0))

19
```
Edit: For posterity, changed links to code to have today's commit instead of 'master'.

**Comment by georgedahl** [comment]
I don't think any consistency with TF apis is intended or promised, but in general i personally prefer a dropout API that supplies the fraction to discard, not the fraction to keep.

**Comment by cboulay** [comment]
TF's nn.dropout previously used a `keep_prob` argument, but that's being deprecated and now `rate` (as 1-keep_prob) is expected for better consistency with Keras; consistency with TF depends on which version of TF we're talking about.

I think the argument name in stax.Dropout should be changed to `keep_prob`, or the behaviour changed to match the implied meaning of "dropout rate".

**Comment by mattjj** [comment]
Thanks for pointing this out, and the super clear exposition. Indeed it looks like we went with a weird convention here.

I agree this is worth revising. We are likely to make bigger revisions to stax soon (and/or make a stax2  and maybe a stax3), so it might be a good idea to batch this change into those bigger changes.

**Comment by hawkinsp** [comment]
We're no longer trying to evolve `stax` (and are working towards removing it from the JAX repository), in favor of much more full-featured neural network libraries like `flax` and `haiku`. Closing because I think it makes the issue moot.
